### PR TITLE
[MRESOLVER-269] [MRESOLVER-275] Trusted checksums source and more compact format backed source

### PR DESCRIPTION
**Pull request by cstamas**
This PR implements several improvement issues:
* introduces "trusted" checksums source (and adapts them for provided transport checksums)
* pushed/moved existing implementation as "trusted" source and introduces "compact" (file) based source
* cleanup re naming: before it was "file" source, now there is "file-sparse" (old) and "file-compact" (new)
* also adds minor "cleanup" in AetherModule (stylistic, renames provider private vars and makes them unmodifiable). This does not affects consumer of this library that use sisu (like Maven).

Reason: existing `ProvidedChecksumSource` is all about _transport_ (see API). Uses transport related classes and is meant -- as it's name and package says -- to provide expected checksums for transport related checks (it provides ChecksumKind#PROVIDED, uses RemoteRepository and transfer related classes). OTOH, there may be requirement to "provide" checksums for operations totally unrelated to transport. Hence, the introduced trusted checksums (using non transport related API) is exactly that, provides checksums for given Artifact (optionally factoring in origin as well in form of  ArtifactRepository). This clearly separates "transport realm" and rest of the things.

Along with existing (moved) source, new trusted checksums implementation added that uses more compact format: a "summary" file that contains list of Artifact IDs and checksum per one line. This format is more VCS friendly, and also easier to handle then sparse directories.

By default, new "trusted" checksum sources _are adapted_ to "provided" checksum sources (see `TrustedToProvidedChecksumsSourceAdapter`), so no functionality loss happens.

Perfomed cleanup around trusted checksum sources as well, old one was in wrong place and wrongly named, dropped it (as it was final class), and now we have two sources:
* `sparse-directory` -- behaves exactly same as dropped one, expects provided checksums in "local repo"-like sparse layout
* `summary-file` -- is the new format, where one file `checksums.${checksumExt}` is expected to contain Artifact ID and checksum for given algorithm per line (separated by space)

Both source are able to be "origin aware" when it factors in origin repository ID as well (so one could get `checksums-central.sha1` with all the known trusted checksums for use).

Sources are DISABLED by default, as even if file is present (check possible only for file-compact) it does not mean user want to use it in every project. Enabling them is possible via usual means (`-D...` or by config in `.mvn` directory to make it per-project persistent). All configuration is sourced from repo system session, no system properties used.

Based on work done in https://github.com/apache/maven-resolver/pull/192

Co-authored-by: @raphw <rafael.wth@gmail.com>

---
https://issues.apache.org/jira/browse/MRESOLVER-275 -- Introduce trusted checksums source
https://issues.apache.org/jira/browse/MRESOLVER-269 -- Allow more compact storage of provided checksums

### CODE REVIEW

**Comment by cstamas** [comment]
This PR obsoletes https://github.com/apache/maven-resolver/pull/192 as it provides 100% same change as in that PR but this PR also cleans up in some areas.

**Comment by raphw** [comment]
This looks really good, thanks!

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/FileTrustedChecksumsSourceSupport.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+import org.eclipse.aether.spi.checksums.TrustedChecksumsSource;
+import org.eclipse.aether.spi.connector.checksum.ChecksumAlgorithmFactory;
+import org.eclipse.aether.util.ConfigUtils;
+import org.eclipse.aether.util.DirectoryUtils;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Support class for implementing {@link TrustedChecksumsSource} backed by local filesystem. It implements basic support
+ * like bqsedir calculation, "enabled" flag and "originAware" flag.
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/FileTrustedChecksumsSourceSupport.java:41]: typo

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/CompactFileTrustedChecksumsSource.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+    @Override
+    protected Map<String, String> performLookup( RepositorySystemSession session,
+                                                 Path basedir,
+                                                 Artifact artifact,
+                                                 ArtifactRepository artifactRepository,
+                                                 List<ChecksumAlgorithmFactory> checksumAlgorithmFactories )
+    {
+        final HashMap<String, String> checksums = new HashMap<>();
+        final ConcurrentHashMap<String, ConcurrentHashMap<String, String>> baseDirProvidedHashes = checksumCache
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/CompactFileTrustedChecksumsSource.java:87]: basedirProvidedChecksums

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/CompactFileTrustedChecksumsSource.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+            }
+        }
+        else
+        {
+            prefix = CHECKSUM_FILE_PREFIX;
+        }
+
+        for ( ChecksumAlgorithmFactory checksumAlgorithmFactory : checksumAlgorithmFactories )
+        {
+            ConcurrentHashMap<String, String> algorithmHashes = baseDirProvidedHashes.computeIfAbsent(
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/CompactFileTrustedChecksumsSource.java:108]: algorithmChecksums

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/CompactFileTrustedChecksumsSource.java [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+            String checksum = algorithmHashes.get( ArtifactIdUtils.toId( artifact ) );
+            if ( checksum != null )
+            {
+                checksums.put( checksumAlgorithmFactory.getName(), checksum );
+            }
+        }
+        return checksums;
+    }
+
+    private ConcurrentHashMap<String, String> loadProvidedHashes( Path checksumFile )
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/CompactFileTrustedChecksumsSource.java:123]: Same here

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/FileTrustedChecksumsSourceSupport.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+ * <p>
+ * The configuration keys supported:
+ * <ul>
+ *     <li><pre>aether.trustedChecksumsSource.${name}.enabled</pre> (boolean) must be explicitly set to "true"
+ *     to become enabled</li>
+ *     <li><pre>aether.trustedChecksumsSource.${name}.basedir</pre> (string, path) directory from where implementation
+ *     can use files. If unset, default value is ".checksums" and is resolved from local repository basedir.</li>
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/FileTrustedChecksumsSourceSupport.java:48]: Docs should state, relative or absolute or both.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/SparseFileTrustedChecksumsSource.java [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+import org.eclipse.aether.spi.io.FileProcessor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Sparse file {@link FileTrustedChecksumsSourceSupport} implementation that use specified directory as base
+ * directory, where it expects artifacts checksums on standard Maven2 "local" layout. This implementation uses Artifact
+ * coordinates solely to form path from baseDir, pretty much as Maven local repository does.
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/SparseFileTrustedChecksumsSource.java:47]: basedir

**Review by michael-o — CHANGES_REQUESTED** [review]
* Since the `LocalPathComposer` is used throughout the docs of the sources should mention that this substructure is also reflected in `.checksums/`
* As for the file-compact. I like the idea. Does it make sense to reuse rather a well-known format like BSD checksum file? See also https://github.com/nicoulaj/checksum-maven-plugin/issues/127

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/FileTrustedChecksumsSourceSupport.java [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+     * Uses common {@link DirectoryUtils} to calculate (but not) create basedir for this implementation. Returns
+     * {@code null} if the calculated basedir does not exist.
+     */
+    private Path getBasedir( RepositorySystemSession session )
+    {
+        try
+        {
+            Path basedir = DirectoryUtils.resolveDirectory(
+                    session, LOCAL_REPO_PREFIX, configPropKey( CONF_NAME_BASEDIR ), false );
+            if ( !Files.isDirectory( basedir ) )
```
[the remark this answers is not in the document]
Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/FileTrustedChecksumsSourceSupport.java:138]: It is for "not exists" case (and create=false), so yes, if exists and is NOT a directlry, DirectoryUtils will throw already. So when here, it either exists (and is directory) or not exists (and isDirectory returns false). IMHO the `Files.exists` is not appropriate here, as would not reflect the real intent. 

TL;DR: if exists and is not directory, code does not even arrive here, but I want to reflect the fact basedir is a directory (if exists) or does not exists.

[the remark this answers is not in the document]
Comment (michael-o) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/FileTrustedChecksumsSourceSupport.java:139]: Alright, makes sense. The developer has to take care of that, not us.

**Comment by michael-o** [comment]
> > * Since the `LocalPathComposer` is used throughout the docs of the sources should mention that this substructure is also reflected in `.checksums/`
> 
> The javadoc of sparse source says:
> 
> ```
>  * Sparse file {@link FileTrustedChecksumsSourceSupport} implementation that use specified directory as base
>  * directory, where it expects artifacts checksums on standard Maven2 "local" layout. This implementation uses Artifact
>  * coordinates solely to form path from basedir, pretty much as Maven local repository does.
> ```

I don't see how this answers my question... ?!!

> > * As for the file-compact. I like the idea. Does it make sense to reuse rather a well-known format like BSD checksum file? See also [nicoulaj/checksum-maven-plugin#127](https://github.com/nicoulaj/checksum-maven-plugin/issues/127)
> 
> IMHO no, as BSD checksum file contains filename, here, we operate with ArtifactId ("file format is artifact ID and checksum separated by space per line"), see https://github.com/apache/maven-resolver/blob/master/maven-resolver-util/src/main/java/org/eclipse/aether/util/artifact/ArtifactIdUtils.java#L45

hm...we basically use our current format, but compile multiple files and key them by artifact id. Agree.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/CompactFileTrustedChecksumsSource.java [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+ * The name of this implementation is "file-compact".
+ *
+ * @see ArtifactIdUtils#toId(Artifact)
+ * @since TBD
+ */
+@Singleton
+@Named( CompactFileTrustedChecksumsSource.NAME )
+public final class CompactFileTrustedChecksumsSource
+        extends FileTrustedChecksumsSourceSupport
+{
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/CompactFileTrustedChecksumsSource.java:67]: This class has one inconsistency: All our meta files have the form: `{metafile}.{ext}` or `{metafile}-{repoId}.{ext}`. This should apply here as well. E.g., `./org/apache/maven/resolver/maven-metadata-local.xml`

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/SparseFileTrustedChecksumsSource.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+        final String artifactPath = localPathComposer.getPathForArtifact( artifact, false );
+        for ( ChecksumAlgorithmFactory checksumAlgorithmFactory : checksumAlgorithmFactories )
+        {
+            Path checksumPath = basedir.resolve(
+                    prefix + artifactPath + "." + checksumAlgorithmFactory.getFileExtension() );
+
+            if ( !Files.isRegularFile( checksumPath ) )
+            {
+                continue;
+            }
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/SparseFileTrustedChecksumsSource.java:112]: I think this deserves a log message. I guess developer provided us nonsense.

File: maven-resolver-impl/src/test/java/org/eclipse/aether/internal/impl/checksum/FileTrustedChecksumsSourceTestSupport.java [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+    protected ChecksumAlgorithmFactory checksumAlgorithmFactory;
+
+    private FileTrustedChecksumsSourceSupport subject;
+
+    @Before
+    public void before() throws Exception
+    {
+        session = TestUtils.newSession();
+        // populate local repository
+        Path baseDir = session.getLocalRepository().getBasedir().toPath()
```
Comment (michael-o) [inline maven-resolver-impl/src/test/java/org/eclipse/aether/internal/impl/checksum/FileTrustedChecksumsSourceTestSupport.java:58]: basedir

**Review by michael-o — CHANGES_REQUESTED** [review]

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/CompactFileTrustedChecksumsSource.java [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+    public static final String NAME = "file-compact";
+
+    private static final String CHECKSUM_FILE_PREFIX = "checksums.";
+
+    private static final String CHECKSUMS_CACHE_KEY = NAME + "-checksums";
+
+    private static final Logger LOGGER = LoggerFactory.getLogger( CompactFileTrustedChecksumsSource.class );
+
+    @Inject
```
[the remark this answers is not in the document]
Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/CompactFileTrustedChecksumsSource.java]: isn;t, is explicit but removed it (as guice will figure)

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/CompactFileTrustedChecksumsSource.java (hunk 8 above)
Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/CompactFileTrustedChecksumsSource.java:66]: made it `checksums-central.sha1`

**Comment by cstamas** [comment]
> I don't see how this answers my question... ?!!

I might missed the real intent of your question then...

**Comment by michael-o** [comment]
Going through again...

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/CompactFileTrustedChecksumsSource.java [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+                            LOGGER.warn( "Checksums file '{}' ignored malformed line '{}'", checksumFile, line );
+                        }
+                    }
+                }
+            }
+        }
+        catch ( NoSuchFileException e )
+        {
+            // ignore, will return empty result
+            LOGGER.debug( "Checksums file '{}' not found", checksumFile );
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/CompactFileTrustedChecksumsSource.java]: This is an antipattern according to Joshua Bloch. Test for existance instead of reyling on such exceptions.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/SparseFileTrustedChecksumsSource.java [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Sparse file {@link FileTrustedChecksumsSourceSupport} implementation that use specified directory as base
+ * directory, where it expects artifacts checksums on standard Maven2 "local" layout. This implementation uses Artifact
+ * coordinates solely to form path from basedir, pretty much as Maven local repository does.
+ * <p>
+ * The source may be configured to be "origin aware", in that case it will factor in origin repository ID as well into
+ * base directory name (for example ".checksums/central/...").
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/SparseFileTrustedChecksumsSource.java:50]: Call me stupid, but why don't we use the `LocalPathPrefixComposer` for this. It will behave identically to `EnhancedLocalRepositoryManager`?

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/FileTrustedChecksumsSourceSupport.java [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+    private static final String CONF_NAME_ENABLED = "enabled";
+
+    private static final String CONF_NAME_BASEDIR = "basedir";
+
+    private static final String CONF_NAME_ORIGIN_AWARE = "originAware";
+
+    /**
+     * Visible for testing.
+     */
+    static final String LOCAL_REPO_PREFIX = ".checksums";
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/FileTrustedChecksumsSourceSupport.java:72]: This should be rather `LOCAL_REPO_PREFIX_DIR` since it stands as a dir name .

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/SparseFileTrustedChecksumsSource.java [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+ * <p>
+ * The name of this implementation is "file-sparse".
+ *
+ * @see LocalPathComposer
+ * @since TBD
+ */
+@Singleton
+@Named( SparseFileTrustedChecksumsSource.NAME )
+public final class SparseFileTrustedChecksumsSource
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/SparseFileTrustedChecksumsSource.java:59]: I know what sparse means here and I think it is used incorrectly. Your understanding: A spare directory tree where only a limited amount of files exist for artifacts required. Meaning of sparse file: https://en.wikipedia.org/wiki/Sparse_file
I think the term is abused.

It should be `SparseDirectoryFileTrustedChecksumsSource` or `SparseDirectoryTrustedChecksumsSource`.

See impl of a sparse dir tree: https://svnbook.red-bean.com/en/1.7/svn.advanced.sparsedirs.html

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/CompactFileTrustedChecksumsSource.java [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+import org.eclipse.aether.repository.ArtifactRepository;
+import org.eclipse.aether.spi.connector.checksum.ChecksumAlgorithmFactory;
+import org.eclipse.aether.util.artifact.ArtifactIdUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Compact file {@link FileTrustedChecksumsSourceSupport} implementation that use specified directory as base
+ * directory, where it expects a "summary" file named as "checksums.${checksumExt}" for each checksum algorithm, and
+ * file format is artifact ID and checksum separated by space per line. The format supports comments "#" (hash) and
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/CompactFileTrustedChecksumsSource.java:48]: Stupid question: If you say yourself that it uses a summary file, why just not call it so? `SummaryFileTrustedChecksumsSource`, DI name `summary-file`

**Review by michael-o — CHANGES_REQUESTED** [review]
Re: the origin aware. We have now `aether.enhancedLocalRepository.splitRemoteRepository`. To make it consistent, why not name it instead of `originalAware` `splitRepository`?
This may be obsolete if the LocalPathPrefixComposer is used.

**Comment by michael-o** [comment]
> > I don't see how this answers my question... ?!!
> 
> I might missed the real intent of your question then...

Forget it, I understand now. No further explanation necessary.

**Comment by michael-o** [comment]
Yet another question. It is worthwhile to enable it by default and add checksums on the fly when they are present?

**Review by michael-o — COMMENTED** [review]
Yet another issue with the `.enabled`: I think this is again inconsistent with our existing boolean properties: https://maven.apache.org/resolver/configuration.html
They use the name as boolean property with a default value. So to apply this appoach `.enabled` just can be dropped and its parent used.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/SparseFileTrustedChecksumsSource.java (hunk 13 above)
Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/SparseFileTrustedChecksumsSource.java:50]: LocalPathPrefixComposer is a component, and in that case we could have it behave same as LRM. In other words, if you "split" LRM you'd need to split this one as well, and other way around. I don't see the two "same".

**Comment by cstamas** [comment]
> Yet another question: It is worthwhile to enable it by default and add checksums on the fly when they are present?

This could be similar as "record" functionality is for remote repository filter feature. Yes, it could be possible, but unsure is it worths it: the whole idea of this is to get checksums for _other trusted source_ than remote...

**Comment by cstamas** [comment]
> Yet another issue with the `.enabled`: I think this is again inconsistent with our existing boolean properties: https://maven.apache.org/resolver/configuration.html They use the name as boolean property with a default value. So to apply this appoach `.enabled` just can be dropped and its parent used.

This is fixed now.

**Comment by michael-o** [comment]
> > Yet another question: It is worthwhile to enable it by default and add checksums on the fly when they are present?
> 
> This could be similar as "record" functionality is for remote repository filter feature. Yes, it could be possible, but unsure is it worth it: the whole idea of this is to get checksums for _other trusted source_ than remote...

ОК, let's put this aside.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/SparseFileTrustedChecksumsSource.java (hunk 13 above)
Comment (michael-o) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/SparseFileTrustedChecksumsSource.java]: Can you rephrase your answer?! I frankly do not understand whether everything under `.checksums/` can or cannot behave like with LRM treewise. This should also make `.originAware` obsolete.

Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/SparseFileTrustedChecksumsSource.java]: It can, but I tend to NOT make it 100% same for several reasons:
* if we'd use LocalPathPrefixComposer, then if you use split LRM you MUST use split (originAware) SparseDirectory as well. So it's both or nothing.
* given we want to distinguish here the ORIGIN of checksums, I would even lessen the "resemblency" to local repo (to not have anyone "tempted" to just copy the hashes from his local repo for example)
* both (non origin aware and origin aware) has it's "pros": one can ship "curated" checksums per project (non-origin aware), or one can ship "curated" checksum per repo

Comment (michael-o) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/SparseFileTrustedChecksumsSource.java]: OK, now I understand it makes sense. Would it at least, for consistency reasons, make sense to rename `.originAware` to `.splitRepository`? This term is already established with its semantics?

Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/SparseFileTrustedChecksumsSource.java]: IMHO nope, partially due that above ("lessen the resemblency to LRM"), and partially as these are "trusted checksums" ONLY (that may or may not have origin noted), while "split repository" is, well, as it's name say a "repository", that again implies layout and contents (artifacts, not checksums). Plus, then we'd have a "repository embedded within repository", as `$LOCAL_REPO/.splitRepository` would imply that, no?

**Review by michael-o — CHANGES_REQUESTED** [review]
This is basically good/done, but https://github.com/apache/maven-resolver/blob/master/src/site/markdown/configuration.md needs to be updated for this.

**Review by michael-o — APPROVED** [review]
All good to merge.

**Comment by jira-importer** [comment]
Resolve #765
